DRILL-4199: Add Support for HBase 1.X

adityakishore opened the pull request:
Highlights of the changes:

* Updated HBase dependencies version to 1.1.3 and 1.1.1-mapr-1602-m7-5.1.0 for default and mapr profiles respectively.
* Replaced the old HBase APIs (HBaseAdmin/HTable) with the new HBase 1.1 APIs (Connection/Admin/Table).
* Added `commons-logging` dependency in the `provided` scope to allow HBase test cluster to come up for Unit tests.
* Relaxed banned dependency rule for `commons-logging` library for `storage-hbase` module alone, in provided scope only.
* Removed the use of many deprecated APIs throughout the modules code.
* Added some missing test to HBase storage plugin's test suit.

All unit tests are green.

jacques-n commented:
The jcl-over-slf4j bridge should be all that is necessary. Is HBase doing something weird that means that doesn't work? Or is that just missing for some reason?

jcl-over-slf4j makes it so we get all logging over logback which helps debug hbase issues (I definitely used it when I was working on resolving the hbase guava version issues)

adityakishore commented:
`jcl-over-slf4j` works as long as the application use only the surface logging APIs.

Unfortunately, HBase uses some [internal classes](https://github.com/apache/hbase/blob/branch-1.3/hbase-server/src/main/java/org/apache/hadoop/hbase/http/HttpRequestLog.java) from the `commons-logging` library which are not available in `jcl-over-slf4j`.

I have verified that the dependency does not traverse to other module and certainly does not get bundled with the binary tarball.

jacques-n commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/config/HBasePersistentStore.java:
```diff
@@ -25,36 +25,40 @@
 
+import com.google.common.collect.Iterators;
+
 public class HBasePersistentStore<V> extends BasePersistentStore<V> {
```
What you think about just removing this set of classes?

adityakishore replied:
Do you mean getting rid of HBase persistent store altogether?

jacques-n replied:
Yes. I've seen two recent patches where people are having to maintain this code. However, I thought we basically tell people not to use it. (That HDFS + ZK is our preferred paradigm.) That paradigm definitely outperforms HBase since we don't have to hit the store as part of every query (since we have listeners).

adityakishore replied:
Okay, I'd like to do that as a separate JIRA once I send a feeler out on the user list, just in case someone is using it.

jacques-n replied:
Sounds great, thanks!

jacques-n commented:
I don't know the new semantics of ConnectionFactory. Drill now has plugin lifecycle events so we should close out an hbase connection if a storage plugin is changed or unregistered. Does that make sense in the context of HBase or does it continue to just use a static pool for connections with no way to close out?

Otherwise, LGTM +1

adityakishore commented:
I'll take a look at the plugin life-cycle model and see if we can tie HBase connection life-cycle to it.

As far as I know HBase follows a per-user connection model so I'll have to test how does that play with Drill impersonation.

adityakishore commented:
I have verified that, with these changes, Drill queries run fine with HBase 0.98 cluster too.

adityakishore commented:
The latest patch, 1684daa5a081e1ae7d6d66cb5ee5e2f88015fc53, takes care of HBase connection life-cycle management within the plugin's life-cycle.

adityakishore commented:
@jacques-n I'll wait for your final review before I merge this.

jacques-n commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseStoragePlugin.java:
```diff
@@ -19,36 +19,75 @@
+            }
+            return connection;
+          }
+          private Connection connection = null;
```
Why is this an instance field and not protected? Since the cache loader is already managing a single instance, it doesn't seem like we need to worry about ensuring a singleton.  Separately, since we're behind a cache, we won't call connection.isAborted(). Should we make sure that the get() checks isAborted() and isClosed()

adityakishore replied:
> Why is this an instance field and not protected?

Did you mean static?

adityakishore commented:
The latest patch introduces a separate class `HBaseConnectionManager` and moves the (HBase) connection life-cycle management responsibility from HBaseStoragePlugin to this class.

Also addressed Jacques' other comment regarding testing for connection close/abort in `getConnection()` method.

Added test which verifies that HBase connection close or even HBase cluster restart does not impact running queries against HBase and does not require Drill cluster restart.

jacques-n commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/config/HBasePersistentStore.java:
```diff
@@ -175,7 +179,7 @@ public boolean hasNext()  {
           }
         } catch (IOException e) {
-          throw new DrillRuntimeException("Caught error while fetching rows from for table:" + Bytes.toString(table.getTableName()), e);
+          throw new DrillRuntimeException("Caught error while fetching rows from for table:" + hbaseTableName, e);
```
Can we use UserException.dataReadError(e) for these?

adityakishore replied:
Will do.

jacques-n commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseConnectionManager.java:
```diff
@@ -0,0 +1,92 @@
+      Connection conn = connectionCache.get(key);
+      if (conn.isAborted()
+          || conn.isClosed()) {
+        connectionCache.invalidate(key);
```
It seems like there is a possible race condition here. If two separate threads both realize a key is bad, the first one may invalidate the key and insert a new one, then return that key. Meanwhile, the second thread will invalidate the new connection from the first thread (and thus close the connection) causing the first threads connection to be closed while in use. I think there is a way to check a value in Guava's model to ensure that it is valid before return. If not, you probably need to do a double-checked locking inside the invalidation (or similar).

adityakishore replied:
Thanks for catching this, not sure how it slipped through my own review :(

Couldn't find a simple way to make Guava do this (validate a cached value before returning), so will change it to do both eviction and load atomically in this code itself.

jacques-n commented on contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseStoragePlugin.java:
```diff
@@ -21,34 +21,36 @@
 
 public class HBaseStoragePlugin extends AbstractStoragePlugin {
-  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(HBaseStoragePlugin.class);
+  private static final HBaseConnectionManager hbaseConnectionManager = HBaseConnectionManager.INSTANCE;
```
Any reason you use static here as opposed to one per storage plugin instance? I suppose this reduces overhead in the case of multiple plugin instances to a single target but generally, I'd be inclined to just have an instance per plugin.

adityakishore replied:
The HBaseConnectionManager is static, and is essentially a connection pool which maintains one HBase connection per storage plugin instance as you recommend.

adityakishore commented:
Addressed Jacques' comments on the previous patch.

jacques-n commented on contrib/storage-hbase/pom.xml:
```diff
@@ -77,6 +77,14 @@
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>commons-logging</groupId>
```
Can't we use jcl-over-slf4j?

jacques-n commented:
One comment about jcl, otherwise lgtm +1

adityakishore commented:
Unfortunately we can't, please see https://github.com/apache/drill/pull/443#issuecomment-200580166

jacques-n commented:
Hahaha!

Well at least I'm consistent :P

adityakishore commented:
The latest patch, 323849e48fed967d43108e5cf85ba150a50f073e, moves the GuavaPatcher class to java-exec module and run the GuavaPatcher at Drillbit start (before the Stopwatch class is loaded somewhere else).

@jacques-n Could you please take a look?

adityakishore commented:
The latest and hopefully the last version of the patch (c2d9959e07f47a09a4a11c250f84f4874b7e1db4), modified the GuavaPatcher in a way to make it non-catastrophic failure to not able to patch Guava.

This allows the JDBC driver to not crash when starting Drill in embedded mode (mostly used in unit tests).
